protocol-spec/generic.md

```diff
@@ -134,5 +142,197 @@
 ]
 ```
 
+## VibrateCmd
+
+**Description:** Causes a toy that supports vibration to run specific vibration motors at a certain speeds. In order to abstract the dynamic range of different toys, the values are sent as a dictionary of vibration motor indexes against floats with a range of \[0.0-1.0\]
+
+**Introduced In Version:** 1
+
+**Message Version:** 1
+
+**Fields:**
+
+* _Id_ \(unsigned int\): Message Id
+* _DeviceIndex_ \(unsigned int\): Index of device
+* _Speeds_ \(array\): Vibration speeds
+  * _Index_ \(unsigned int\): Index of vibration motor
+  * _Speed_ \(double\): Vibration speed with a range of \[0.0-1.0\]
+
+**Expected Response:**
+
+* Ok message with matching Id on successful request.
+* Error message on value or message error.
+
+**Flow Diagram:**
+
+**Serialization Example:**
+
+```json
+[
+  {
+    "VibrateCmd": {
+      "Id": 1,
+      "DeviceIndex": 0,
+      "Speeds": [
+        {
+          "Index": 0,
+          "Speed": 0.5
+        },
+        {
+          "Index": 1,
+          "Speed": 1
+        }
+      ]
+    }
+  }
+]
+```
+
+## LinearCmd
+
+**Description:** Causes a toy that supports linear movement to reposition it's linear actuators at a certain speeds. In order to abstract the dynamic ranges (both speed and movement) of different toys, the values are sent as a dictionary of linear actuator indexes against objects encapsulating floats with a range of \[0.0-1.0\] for speed and position.
+
+**Introduced In Version:** 1
+
+**Message Version:** 1
+
+**Fields:**
+
+* _Id_ \(unsigned int\): Message Id
+* _DeviceIndex_ \(unsigned int\): Index of device
+* _Vectors_ \(array\): Linear actuator speeds and positions
+  * _Index_ \(unsigned int\): Index of linear actuator
+  * _Speed_ \(double\): Movement speed with a range of \[0.0-1.0\]
+  * _Position_ \(double\): Target position with a range of \[0.0-1.0\]
+
+**Expected Response:**
+
+* Ok message with matching Id on successful request.
+* Error message on value or message error.
+
+**Flow Diagram:**
+
+
+**Serialization Example:**
+
+```json
+[
+  {
+    "LinearCmd": {
+      "Id": 1,
+      "DeviceIndex": 0,
+      "Vectors": [
+        {
+          "Index": 0,
+          "Speed": 0.5,
+          "Position": 0.3
+        },
+        {
+          "Index": 1,
+          "Speed": 1,
+          "Posiion": 0.8
+        }
+      ]
+    }
+  }
+]
+```
+
+
+
+## RotateCmd
+
+**Description:** Causes a toy that supports rotation to run specific rotation motors at a certain speeds in specified directions. In order to abstract the dynamic range of different toys, the values are sent as a dictionary of rotation motor indexes against objects encapsulating the speed as a float with a range of \[0.0-1.0\] and the direction as a boolean (true being clockwise). **Note:** clockwise may be subjective.
```

Comment:
Why would clockwise be subjective?

Reply:
Because some devices don't actually let you set it, just change the direction (Nora).
Also, is clockwise from the perspective of the user always the same direction?
I'm also not sure which way around you'd hold the device to tell the direction: are you holding the Nora by the handle with the tip facing towards or away from you. Does it make a difference if the Nora is behind you?

Reply:
Wow. Ok. We need some sort of hand rule for toy direction specification I guess?

Reply:
It should probably be "point the part of the device that interacts with your body at your face, the part that moves is either going clockwise or anticlockwise from this perspective".
The gotcha is devices that don't let you specify the direction, but let you reverse it: they'll respond in a way close enough to what you'd expect (I bet it'd feeling the same), but clockwise might be too absolute a term